chore: upgrade eslint and typescript-eslint

iosh opened the pull request:
<!--
Thank you for using Hardhat and taking the time to send a Pull Request!

If you are introducing a new feature, please discuss it in an Issue or with someone from the team before submitting your change.

Please:
 - consider the checklist items below
 - keep the ones that make sense for your PR, and
 - DELETE the items that DON'T make sense for your PR.

## Note about small PRs and airdrop farming

We generally really appreciate external contributions, and strongly encourage meaningful additions and fixes! However, due to a recent increase in small PRs potentially created to farm airdrops, we might need to close a PR without explanation if any of the following apply:

- It is a change of very minor value that still requires additional review time/fixes (e.g. PRs fixing trivial spelling errors that can’t be merged in less than a couple of minutes due to incorrect suggestions)
- It introduces inconsequential changes (e.g. rewording phrases)
- The author of the PR does not respond in a timely manner
- We suspect the Github account of the author was created for airdrop farming
-->

- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [x] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [] I didn't do anything of this.

---

<!-- Add a description of your PR here -->

## ESLint from v8 to v9

resolve #5568

This PR migrates the project's ESLint setup from version 8 to version 9. This includes adopting the new flat configuration format (`eslint.config.cjs`) and updating related dependencies.

### Key Changes:

1.  **New Configuration Package (`config`) in v-next:**

    - Created a new internal package (`config`) to hold the shared ESLint 9 configuration.
    - Moved ESLint-related dependencies (plugins, parser, resolvers) into this package.
    - Removed these dependencies from individual workspace packages to centralize configuration management.

2.  **Dependency Updates:**

    - `eslint` -> `9.25.1`
    - `typescript-eslint` -> `8.31.0`
    - `eslint-plugin-import` -> `v2.31.0`
    - `eslint-plugin-no-only-tests` -> `3.3.0`

3.  **Configuration File Renaming:**

    - Renamed `.eslintrc.cjs` to `eslint.config.cjs` across the entire workspace, including the root and individual packages.

4.  **Configuration Syntax Updates (ESLint 9 Flat Config):**

    - Replaced `env` with `languageOptions`. (See [ESLint Docs](https://eslint.org/docs/latest/use/configure/migration-guide#configuring-language-options))
    - Replaced the deprecated `@typescript-eslint/ban-types` rule with the recommended `@typescript-eslint/no-wrapper-object-types`. (See [TS-ESLint Blog](https://typescript-eslint.io/blog/announcing-typescript-eslint-v8-beta/#replacement-of-ban-types))
    - Updated the `plugins` configuration from an array (`plugins: ["import"]`) to an object format (`plugins: { import: eslintPluginImport }`). (See [ESLint Docs](https://eslint.org/docs/latest/use/configure/plugins#plugin-configuration))
    - Adapted `overrides` config . (See [ESLint Docs](https://eslint.org/docs/latest/use/configure/migration-guide#glob-based-configs))
    - The internal `createConfig` helper function now returns an array of configuration objects (`import("eslint").Linter.Config[]`) instead of a single object.

5.  **Workspace Package Updates:**

    - Updated ESLint config file names (`.eslintrc.cjs` -> `eslint.config.cjs`).
    - Removed the following dependencies from individual packages (now managed centrally in `config-v-next`):

      - `@eslint-community/eslint-plugin-eslint-comments`
      - `@typescript-eslint/eslint-plugin`
      - `@typescript-eslint/parser`
      - `eslint-import-resolver-typescript`
      - `eslint-plugin-import`
      - `eslint-plugin-no-only-tests`
      - `typescript-eslint`

      - `eslint-config-prettier` (unused)

I have run `pnpm eslint` in each package for both the v8 and v9 configurations. When the number of errors/warnings differed, `npx eslint --print-config` was used to compare the effective configurations and check the rule is same.

~~  Update 2025-4-20 ~~

- ~~Instead of `pnpm eslint` run `npx eslint --rule 'no-restricted-imports: off' "src/**/*.ts" "test/**/*.ts"`~~
- ~~ignore `"error  'node:test' import is restricted from being used by a pattern. Use the 'node:' prefix to import built-in Node.js modules  no-restricted-imports"` in the results~~

** Update 2025-4-25: **
Fix eslint check error and warnings:

- **hardhat:**

  - @typescript-eslint/switch-exhaustiveness-check: 2
    Disable errors:

    - Ignore Cases not matched: undefined
    - Ignore other platforms

  - @typescript-eslint/no-unused-vars: 8

    - Added the configuation `caughtErrorsIgnorePattern: "^_",` and rename the unused `error` to `_error`

  - Unused eslint-disable directive: 5 (warnings)
    - Remove all Unused eslint-disable directive

- **hardhat-errors:**

  - v9: 0 error, 0 warnings
    - No changes

- **hardhat-ethers:**

  - v9: 0 error, 0 warnings
    - No changes

- **hardhat-ethers-chai-matchers:**

  - v9: 4 errors, 1 warning
    - @typescript-eslint/no-unused-vars: 4
      - Rename the unused `error` to `_error`
    - Unused eslint-disable directive: 1 (warnings)
      - Remove all Unused eslint-disable directive

- **hardhat-ignition:**

  - v9: 0 errors, 2 warnings
    - Unused eslint-disable directive (warnings)
      - Remove all Unused eslint-disable directive

- **hardhat-ignition-core:**

  - v9: 8 errors, 0 warnings

    - import/order: 4

      - Fix import order

    - @typescript-eslint/strict-boolean-expressions: 1

      - Add undefined check

    - @typescript-eslint/no-unused-vars: 1

      - Rename the unused `error` to `_error`

    - @typescript-eslint/switch-exhaustiveness-check: 2
      - Disable lint

- **hardhat-ignition-ethers:**

  - v9: 0 problems
    - No changes

- **hardhat-ignition-ui:**

  - v9: 0 errors, 2 warnings
    - Unused eslint-disable directive: 2 (warnings)
      - Remove all Unused eslint-disable directive

- **hardhat-ignition-viem:**

  - v9: 19 errors, 0 warnings
    - @typescript-eslint/ban-ts-comment: 17
      - Add `Expect an error` to comment
    - @typescript-eslint/strict-boolean-expressions: 2
      - Add `This function return boolean` to comment

- **hardhat-keystore:**

  - v9: 0 errors, 1 warning
    - Remove all Unused eslint-disable directive

- **hardhat-mocha:**

  - v9: 0 error, 0 warnings

- **hardhat-network-helpers:**

  - v9: 1 error, 0 warnings
    - @typescript-eslint/strict-boolean-expressions: 1
      - Add `=== true` check

- **hardhat-node-test-reporter:**

  - v9: 1 errors, 0 warnings
    - @typescript-eslint/switch-exhaustiveness-check： 1

- **hardhat-node-test-runner:**

  - v9: 0 error, 0 warnings

- **hardhat-test-utils:**

  - v9: 3 errors, 0 warnings
    - @typescript-eslint/no-unused-vars : 3
      - Rename the unused `error` to `_error`

- **hardhat-typechain:**

  - v9: 0 errors, 5 warnings
    - Unused eslint-disable directive: 5 (warnings)
      - Remove all Unused eslint-disable directive

- **hardhat-utils:**

  - v9: 6 errors, 0 warnings

    - @typescript-eslint/switch-exhaustiveness-check: 1

      - Disable and add comment

    - @typescript-eslint/no-unused-vars: 5
      - Remove all Unused eslint-disable directive

- **hardhat-viem:**

  - v8: 1 error, 0 warnings
    - @typescript-eslint/strict-boolean-expressions: 1
  - v9: 0 errors, 0 warnings

- **hardhat-zod-utils:**

  - v8: 2 errors, 0 warnings
    - @typescript-eslint/strict-boolean-expressions: 2
  - v9: 0 error, 0 warnings

- **template-package:**

  - v8: 2 errors, 0 warnings
  - v9: 2 errors, 0 warnings

- **hardhat-toolbox-mocha-ethers**

  - v8: 1 errors, 0 warnings
  - v9: 1 errors, 0 warnings

-  hardhat-toolbox-viem
  - v8: 1 errors, 0 warnings
  - v9: 1 errors, 0 warnings

iosh commented:
The eslint-config-prettier plugin was not being used, so it was removed.

schaable commented on v-next/hardhat-ignition-core/package.json:
```diff
@@ -58,24 +57,17 @@
-    "eslint-import-resolver-typescript": "^3.6.1",
-    "eslint-plugin-import": "2.29.1",
+    "eslint": "^9.23.0",
     "eslint-plugin-mocha": "^10.0.0",
```
is this plugin used?

iosh replied:
not in used.

schaable commented on v-next/hardhat-ignition-ethers/eslint.config.cjs:
```diff
@@ -0,0 +1,14 @@
+
+const configs = createConfig(__filename);
+
+/**
+ * @type {import("eslint").Linter.Config}
+ */
+const overrideConfig = {
+  ignores: ["/test/fixture-projects"],
+};
```
Let’s just add this to the base config, ignoring `/test/**/fixture-projects`.

schaable commented on v-next/hardhat-ignition-ui/package.json:
```diff
@@ -27,15 +26,10 @@
-    "eslint-plugin-no-only-tests": "3.1.0",
+    "eslint": "^9.23.0",
     "eslint-plugin-mocha": "^10.0.0",
     "eslint-plugin-react-hooks": "^5.1.0",
     "eslint-plugin-react-refresh": "^0.4.18",
```
are these plugins used?

iosh replied:
not in used.

schaable commented on v-next/hardhat-ignition-ethers/package.json:
```diff
@@ -43,26 +43,19 @@
-    "eslint-import-resolver-typescript": "^3.6.1",
-    "eslint-plugin-import": "2.29.1",
+    "eslint": "^9.23.0",
     "eslint-plugin-mocha": "^10.0.0",
```
is this plugin used?

schaable commented on v-next/hardhat-ignition-viem/eslint.config.cjs:
```diff
@@ -0,0 +1,16 @@
+ * @type {import("eslint").Linter.Config}
+ */
+const overrideConfig = {
+  ignores: ['/test/fixture-projects'],
```
remember to remove this line once you moved the ignore pattern to the base config

schaable commented on v-next/hardhat-ignition-viem/package.json:
```diff
@@ -44,25 +44,18 @@
-    "eslint-import-resolver-typescript": "^3.6.1",
-    "eslint-plugin-import": "2.29.1",
+    "eslint": "^9.23.0",
     "eslint-plugin-mocha": "^10.0.0",
```
same question

schaable commented on v-next/hardhat-ignition/package.json:
```diff
@@ -59,26 +59,18 @@
-    "eslint-plugin-import": "2.29.1",
-    "eslint-plugin-no-only-tests": "3.1.0",
+    "eslint": "^9.23.0",
     "eslint-plugin-mocha": "^10.0.0",
```
same question

schaable commented on v-next/hardhat-node-test-reporter/eslint.config.cjs:
```diff
@@ -1,8 +1,11 @@
+ * * @type {import("eslint").Linter.Config}
+ */
+const overrideConfig = {
   files: ["integration-tests/**/*.ts"],
```
I’m thinking we should add this path to the base test config so it can be shared across all packages, without needing to add this override in each one.

schaable commented on v-next/hardhat-test-utils/eslint.config.cjs:
```diff
@@ -0,0 +1,13 @@
+
+const configs = createConfig(__filename, { enforceHardhatTestUtils: false });
+
+/**
+ * * @type {import("eslint").Linter.Config}
+ */
+const overrideConfig = {
+  files: ["src/**/*.ts"],
+};
+
+configs.push(overrideConfig);
```
why is this override needed?

iosh replied:
Add to base config.

schaable commented on v-next/hardhat-utils/eslint.config.cjs:
```diff
@@ -0,0 +1,14 @@
+
+const configs = createConfig(__filename, { onlyHardhatError: false });
+
+/**
+ * * @type {import("eslint").Linter.Config}
+ */
+const overrideConfig = {
+  files: ["src/**/*.ts"],
+};
+
+configs.push(overrideConfig);
```
why is this override needed?

schaable commented on v-next/hardhat/package.json:
```diff
@@ -64,27 +64,18 @@
     "expect-type": "^0.19.0",
     "prettier": "3.2.5",
-    "rimraf": "^5.0.5",
-    "typescript": "~5.5.0",
```
I think you accidentally deleted `typescript`.

schaable reviewed (changes requested):
Hi @iosh! Awesome work! I left some comments and questions.

iosh commented:
I saw some test tailed. let me check it.

schaable commented:
> I saw some test tailed. let me check it.

It looks like it's due to diverging versions of the eslint plugin. While you're at it, do you mind:
 - Pinning the versions of eslint and its plugins to avoid unexpected linting errors from patch version differences between local and CI
 - Bumping all eslint-related dependencies to their latest versions

iosh commented:
> > I saw some test tailed. let me check it.
> 
> It looks like it's due to diverging versions of the eslint plugin. While you're at it, do you mind:
> 
> * Pinning the versions of eslint and its plugins to avoid unexpected linting errors from patch version differences between local and CI
> * Bumping all eslint-related dependencies to their latest versions

OK. and some eslint check failed, I need check why.

iosh commented:
@schaable Hi, I have fixed the ESLint errors. Although there were a lot of changes, they were simple. So, I only made one commit and updated the PR description.

iosh commented:
The [[config-v-next] lint](https://github.com/NomicFoundation/hardhat/actions/runs/14665456908/job/41163011013#logs) job failed because the config-v-next package was not in the v-next folder. I have moved the package into the v-next folder and updated the relevant references to fix this.


I tried multiple times to reproduce the error from CI locally bug couldn't, which left me puzzled. After numerous tries, i finally realized it was because i hadn't clear the build cache and rebuild the code.

schaable commented:
@iosh we can ignore the `config` project in `scripts/check-v-next-npm-scripts.js`

iosh commented on v-next/config/package.json:
```diff
@@ -0,0 +1,35 @@
+  "devDependencies": {
+    "@eslint-community/eslint-plugin-eslint-comments": "4.5.0",
+    "@types/eslint": "9.6.1",
+    "@types/debug": "^4.1.7",
```
Removed.
